src/test/components/EmptyThreadIndicator.test.js

```diff
@@ -0,0 +1,215 @@
+
+// @flow
+jest.mock('react-dom');
+
+import * as React from 'react';
+import EmptyThreadIndicator, {
+  getIndicatorPositions,
+} from '../../components/header/EmptyThreadIndicator';
+import renderer from 'react-test-renderer';
+import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';
+import { getBoundingBox } from '../fixtures/utils';
+import type { StartEndRange } from '../../types/units';
+import ReactDOM from 'react-dom';
+
+describe('EmptyThreadIndicator', function() {
+  // Sizing for the containing dom node.
+  const width = 100;
+  const height = 10;
+
+  ReactDOM.findDOMNode = jest.fn(() => {
+    // findDOMNode uses nominal typing instead of structural (null | Element | Text), so
+    // opt out of the type checker for this mock by returning `any`.
+    const mockEl = ({
+      getBoundingClientRect: () => getBoundingBox(width, height),
+    }: any);
+    return mockEl;
+  });
+
+  const { profile } = getProfileFromTextSamples(`
+    A A A
+  `);
+
+  const thread = profile.threads[0];
+  thread.processStartupTime = 2;
+  thread.processShutdownTime = 10;
+  thread.registerTime = 3;
+  thread.unregisterTime = 9;
+  thread.samples.time = [5, 6, 7];
+
+  // Make it really easy to generate the component's props.
+  function propsFromViewportRange(viewport: StartEndRange) {
+    return {
+      rangeStart: viewport.start,
+      rangeEnd: viewport.end,
+      thread: thread,
+      interval: 1,
+      unfilteredSamplesRange: { start: 5, end: 8 },
+      width,
+      height: 10,
+    };
+  }
+  describe('rendering', function() {
+    it('matches the snapshot when rendering all three types of indicators', () => {
+      const props = propsFromViewportRange({ start: 0, end: 10 });
+      expect(
+        renderer
+          .create(
+            // The props have to be passed in manually to avoid adding the SizeProps.
+            <EmptyThreadIndicator
+              rangeStart={props.rangeStart}
+              rangeEnd={props.rangeEnd}
+              thread={props.thread}
+              interval={props.interval}
+              unfilteredSamplesRange={props.unfilteredSamplesRange}
+            />
+          )
+          .toJSON()
+      ).toMatchSnapshot();
+    });
+
+    it('matches the snapshot when rendering no indicators', () => {
+      // This range has samples throughout it.
+      const props = propsFromViewportRange({ start: 5.5, end: 6.5 });
+      expect(
+        renderer
+          .create(
+            // The props have to be passed in manually to avoid adding the SizeProps.
+            <EmptyThreadIndicator
+              rangeStart={props.rangeStart}
+              rangeEnd={props.rangeEnd}
+              thread={props.thread}
+              interval={props.interval}
+              unfilteredSamplesRange={props.unfilteredSamplesRange}
+            />
+          )
+          .toJSON()
+      ).toMatchSnapshot();
+    });
+  });
+
+  describe('startup empty thread indicator', function() {
```

Comment:
All of this behavior scared me because it's kind of magical, so I tried to rigorously test it.